### ISSUE TITLE
Database drivers missing (regardless of Java version used)

### ISSUE THREAD

**Issue by arabold** [metabase/metabase#6486]
For some reason it seems Metabase isn't loading any database drivers anymore other than for the already configured databases. When I try to add a new database in the admin panel the list shows only "MySQL" and "Postgres" (which I have configured in an earlier version of Metabase). Everything else is missing; in particular I wanted to add a Redshift database.

I've found issue #2881 which _seems_ similar, but I've tested using both `java-1.7.0` and `java-1.8.0` packages without luck.

My system is an Amazon Linux EC2 instance.

Calling `java -version` shows this:
```
openjdk version "1.8.0_151"
OpenJDK Runtime Environment (build 1.8.0_151-b12)
OpenJDK 64-Bit Server VM (build 25.151-b12, mixed mode)
```

Im starting Metabase using Postgres as configuration database (`MB_DB_TYPE=postgres`, etc.) and with encryption enabled (`MB_ENCRYPTION_SECRET_KET=...`) as a non-root user on port 8080.

Metabase starts up fine and shows no errors in the log even when running with `log4j.logger.metabase=DEBUG`. For some reason it only seems to load the `postgres` driver:
```
11-28 15:50:54 INFO metabase.util :: Loading Metabase...
11-28 15:51:02 INFO util.encryption :: DB details encryption is ENABLED for this Metabase instance. 🔐 
See http://www.metabase.com/docs/latest/operations-guide/start.html#encrypting-your-database-connection-details-at-rest for more information.
11-28 15:51:15 INFO metabase.core :: Starting Metabase in STANDALONE mode
11-28 15:51:15 INFO metabase.core :: Launching Embedded Jetty Webserver with config:
 {:port 80}

11-28 15:51:15 INFO metabase.core :: Starting Metabase version v0.26.2 (3b65f11 release-0.26.2) ...
11-28 15:51:15 INFO metabase.core :: System timezone is 'Zulu' ...
11-28 15:51:15 INFO metabase.core :: Setting up and migrating Metabase DB. Please sit tight, this may take a minute...
11-28 15:51:15 INFO metabase.db :: Verifying postgres Database Connection ...
11-28 15:51:15 DEBUG metabase.driver :: Registered driver :postgres 🚚
11-28 15:51:17 INFO metabase.db :: Verify Database Connection ...  ✅
11-28 15:51:17 INFO metabase.db :: Running Database Migrations...
11-28 15:51:17 INFO metabase.db :: Setting up Liquibase...
11-28 15:51:17 INFO metabase.db :: Liquibase is ready.
11-28 15:51:17 INFO metabase.db :: Checking if Database has unrun migrations...
11-28 15:51:20 INFO metabase.db :: Database Migrations Current ...  ✅
com.mchange.v2.cfg.DelayedLogItem [ level -> FINE, text -> "The configuration file for resource identifier 'hocon:/reference,/application,/c3p0,/' could not be found. Skipping.", exception -> null]
11-28 15:51:21 INFO db.migrations :: Running all necessary data migrations, this may take a minute.
11-28 15:51:21 INFO db.migrations :: Finished running data migrations.
11-28 15:51:21 DEBUG metabase.task :: Starting Quartz Scheduler
11-28 15:51:21 INFO metabase.metabot :: Starting MetaBot WebSocket monitor thread...
11-28 15:51:21 DEBUG metabase.metabot :: MetaBot WebSocket is closed. Reconnecting now.
11-28 15:51:21 INFO metabase.core :: Metabase Initialization COMPLETE
```

There is no console output in the browser window (Chrome).

Any suggestions?

**Comment by camsaul** [comment]
How are you running Metabase? With the JAR?

**Comment by camsaul** [comment]
When did this start happening?

**Comment by arabold** [comment]
Yes, I'm running `java -jar metabase.jar`. In fact, I'm using an `init.d` script that sets up everything.

```
#!/bin/sh
export MB_DB_TYPE=postgres
export MB_DB_DBNAME=metabase
export MB_DB_PORT=5432
export MB_DB_USER=metabase
export MB_DB_PASS='m=E\d#&E&6UC"$5twu$BKb$KbHL!y(qd'
export MB_DB_HOST=stashimi-dw.cwdxqlh90q2j.us-east-1.rds.amazonaws.com
export MB_JETTY_PORT=8080
export MB_ENCRYPTION_SECRET_KEY='TUSdND8Wk+slrBHBKvQBbX+M11rO7GE1wgqpQSX+KRY='

SERVICE_NAME=MetaBase
JAVA_BIN=/usr/bin/java
PATH_TO_JAR=/opt/metabase
PATH_TO_LOG=/var/log
PID_PATH_NAME=/home/metabase/.metabase.pid
case $1 in
    start)
        echo "Starting $SERVICE_NAME ..."
        touch $PATH_TO_LOG
        if [ ! -f $PID_PATH_NAME ]; then
            su metabase -c "nohup $JAVA_BIN -Dlog4j.configuration=file:$PATH_TO_JAR/log4j.properties -jar $PATH_TO_JAR/metabase >$PATH_TO_LOG/metabase.log 2>&1" & echo $! > $PID_PATH_NAME
            echo "$SERVICE_NAME started ..."
        else
            echo "$SERVICE_NAME is already running ..."
        fi
    ;;
#...
```

Unfortunately I cannot tell exactly when this happened. I noticed today when I tried to add Redshift. Before that I've only added Postgres and those worked as Postgres is still in the list. 

I'm going to try different versions locally now.

**Comment by arabold** [comment]
Mhm... I've started the same `metabase.jar` locally using H2 instead of Postgres. So basically similar to a clean installation _but on the very same machine_. Now it loads all drivers on startup:

```
java -Dlog4j.configuration=file:log4j.properties -jar metabase-0.26.2.jar 
11-28 19:19:48 INFO metabase.util :: Loading Metabase...
11-28 19:19:57 INFO util.encryption :: DB details encryption is DISABLED for this Metabase instance. 🔓 
See http://www.metabase.com/docs/latest/operations-guide/start.html#encrypting-your-database-connection-details-at-rest for more information.
11-28 19:20:10 INFO metabase.core :: Starting Metabase in STANDALONE mode
11-28 19:20:10 INFO metabase.core :: Launching Embedded Jetty Webserver with config:
 {:port 8081}

11-28 19:20:11 INFO metabase.core :: Starting Metabase version v0.26.2 (3b65f11 release-0.26.2) ...
11-28 19:20:11 INFO metabase.core :: System timezone is 'Zulu' ...
11-28 19:20:12 DEBUG metabase.driver :: Registered driver :bigquery 🚚
11-28 19:20:12 DEBUG metabase.driver :: Registered driver :crate 🚚
11-28 19:20:12 DEBUG metabase.driver :: Registered driver :druid 🚚
11-28 19:20:12 WARN metabase.driver :: No -init-driver function found for 'metabase.driver.google'
11-28 19:20:12 DEBUG metabase.driver :: Registered driver :googleanalytics 🚚
11-28 19:20:12 DEBUG metabase.driver :: Registered driver :h2 🚚
11-28 19:20:12 DEBUG metabase.driver :: Registered driver :mongo 🚚
11-28 19:20:12 DEBUG metabase.driver :: Registered driver :mysql 🚚
11-28 19:20:13 DEBUG metabase.driver :: Registered driver :postgres 🚚
11-28 19:20:13 DEBUG metabase.driver :: Registered driver :presto 🚚
11-28 19:20:13 DEBUG metabase.driver :: Registered driver :redshift 🚚
11-28 19:20:13 DEBUG metabase.driver :: Registered driver :sqlite 🚚
11-28 19:20:13 DEBUG metabase.driver :: Registered driver :sqlserver 🚚
11-28 19:20:13 INFO metabase.core :: Setting up and migrating Metabase DB. Please sit tight, this may take a minute...
11-28 19:20:13 INFO metabase.db :: Verifying h2 Database Connection ...
11-28 19:20:14 INFO metabase.db :: Verify Database Connection ...  ✅
11-28 19:20:14 INFO metabase.db :: Running Database Migrations...
11-28 19:20:14 INFO metabase.db :: Setting up Liquibase...
11-28 19:20:14 INFO metabase.db :: Liquibase is ready.
11-28 19:20:14 INFO metabase.db :: Checking if Database has unrun migrations...
11-28 19:20:19 INFO metabase.db :: Database has unrun migrations. Waiting for migration lock to be cleared...
11-28 19:20:19 INFO metabase.db :: Migration lock is cleared. Running migrations...
11-28 19:21:09 INFO metabase.db :: Database Migrations Current ...  ✅
com.mchange.v2.cfg.DelayedLogItem [ level -> FINE, text -> "The configuration file for resource identifier 'hocon:/reference,/application,/c3p0,/' could not be found. Skipping.", exception -> null]
11-28 19:21:10 INFO db.migrations :: Running all necessary data migrations, this may take a minute.
11-28 19:21:10 INFO db.migrations :: Running data migration 'set-card-database-and-table-ids'...
11-28 19:21:10 INFO db.migrations :: Running data migration 'set-mongodb-databases-ssl-false'...
11-28 19:21:10 INFO db.migrations :: Running data migration 'set-default-schemas'...
11-28 19:21:10 INFO db.migrations :: Running data migration 'set-admin-email'...
11-28 19:21:10 INFO db.migrations :: Running data migration 'remove-database-sync-activity-entries'...
11-28 19:21:10 INFO db.migrations :: Running data migration 'update-dashboards-to-new-grid'...
11-28 19:21:10 INFO db.migrations :: Running data migration 'migrate-field-visibility-type'...
11-28 19:21:10 INFO db.migrations :: Running data migration 'add-users-to-default-permissions-groups'...
11-28 19:21:10 INFO models.permissions-group :: Created magic permissions group 'All Users' (ID = 1)
11-28 19:21:10 INFO models.permissions-group :: Created magic permissions group 'Administrators' (ID = 2)
11-28 19:21:10 INFO db.migrations :: Running data migration 'add-admin-group-root-entry'...
11-28 19:21:10 DEBUG models.permissions :: Granting permissions for group 2: /
11-28 19:21:10 INFO db.migrations :: Running data migration 'add-databases-to-magic-permissions-groups'...
11-28 19:21:10 INFO models.permissions-group :: Created magic permissions group 'MetaBot' (ID = 3)
11-28 19:21:10 INFO db.migrations :: Running data migration 'migrate-field-types'...
11-28 19:21:10 INFO db.migrations :: Running data migration 'fix-invalid-field-types'...
11-28 19:21:10 INFO db.migrations :: Running data migration 'copy-site-url-setting-and-remove-trailing-slashes'...
11-28 19:21:10 INFO db.migrations :: Running data migration 'migrate-query-executions'...
11-28 19:21:10 INFO db.migrations :: Running data migration 'drop-old-query-execution-table'...
11-28 19:21:10 INFO db.migrations :: Running data migration 'ensure-protocol-specified-in-site-url'...
11-28 19:21:10 INFO db.migrations :: Finished running data migrations.
11-28 19:21:10 INFO metabase.events :: Starting events listener: metabase.events.activity-feed 👂
11-28 19:21:10 INFO metabase.events :: Starting events listener: metabase.events.dependencies 👂
11-28 19:21:10 INFO metabase.events :: Starting events listener: metabase.events.driver-notifications 👂
11-28 19:21:10 INFO metabase.events :: Starting events listener: metabase.events.last-login 👂
11-28 19:21:10 INFO metabase.events :: Starting events listener: metabase.events.metabot-lifecycle 👂
11-28 19:21:10 INFO metabase.events :: Starting events listener: metabase.events.notifications 👂
11-28 19:21:10 INFO metabase.events :: Starting events listener: metabase.events.revision 👂
11-28 19:21:10 INFO metabase.events :: Starting events listener: metabase.events.sync-database 👂
11-28 19:21:10 INFO metabase.events :: Starting events listener: metabase.events.view-log 👂
11-28 19:21:10 DEBUG metabase.task :: Starting Quartz Scheduler
11-28 19:21:10 INFO metabase.task :: Loading tasks namespace: metabase.task.follow-up-emails 📆
11-28 19:21:10 INFO metabase.task :: Loading tasks namespace: metabase.task.send-anonymous-stats 📆
11-28 19:21:10 INFO metabase.task :: Loading tasks namespace: metabase.task.send-pulses 📆
11-28 19:21:10 INFO metabase.task :: Loading tasks namespace: metabase.task.sync-databases 📆
11-28 19:21:11 INFO metabase.task :: Loading tasks namespace: metabase.task.upgrade-checks 📆
11-28 19:21:11 INFO metabase.core :: Looks like this is a new installation ... preparing setup wizard
11-28 19:21:11 INFO metabase.core :: Please use the following url to setup your Metabase installation:

http://localhost:8081/setup/

```

**Comment by arabold** [comment]
Okay, I think I got the issue! It has nothing to do with the configuration but actually how I start the jar.

I'm using a symlink `metabase -> metabase-0.26.2.jar` and then launch `java -jar metabase` instead of `java -jar metabase-0.26.2.jar`. 

If I launch the symlink, the drivers don't load. If I launch the actual jar, it works as expected.

Is this supposed to work or expected?

**Comment by camsaul** [comment]
That's strange. It loaded the drivers dynamically when it launches by introspecting the classpath and looking for driver namespaces so it must not be finding them for some reason when the JAR is symlinked

**Comment by BlankRain** [comment]
interesting ...

**Comment by camsaul** [comment]
This is fixed by the new custom classloader stuff in 0.32.0 and verified working on Java 8-11
